### ISSUE TITLE
Err needs subclasses

### ISSUE THREAD

**Issue by paul-hammant** [jooby-project/jooby#1243]
File MutantImpl:

```
  @Override
  public <T> T to(final TypeLiteral<T> type, final MediaType mtype) {
    T result = (T) results.get(type);
    if (result == null) {
      try {
        result = parser.convert(type, mtype, data);
        if (result == ParserExecutor.NO_PARSER) {
          Object[] md = md();
          throw new Err((Status) md[2], String.format(FAILURE, md[1], type));
        }
        results.put(type, result);
      } catch (NoSuchElementException ex) {
        Object[] md = md();
        throw new Err.Missing(String.format(REQUIRED, md[1]));
      } catch (Err ex) {
        throw ex;
      } catch (Throwable ex) {
        Object[] md = md();
        throw new Err(parser.statusCode(ex), String.format(FAILURE, md[1], type), ex);
      }
    }
    return result;
  }
```

The problem is that `throw new Err(...);` is vague. The  logic in handler methods may want to differentiate between bad-mime-type and could-not-parse-json-into-pojo.  It'd be great to go even further and know unknown-json-field, bad-json-field-value, and missing-json-field. Of course some of that's missing in Gson too, but you can dream.

**Comment by jknack** [comment]
Probably a ParseException or similar, but nothing more. This is bc we can't guess any possible error outcome from external library (Jackson, Gson, etc...)

Also, there are two specialized subclass:

- Err.BadMediaType
- Err.Missing

**Comment by paul-hammant** [comment]
Could you visit the Jackson/Gson wrapper class you've made?  `public void throwJoobyErrorDerivative(result);`  Mean's you've avoided the if/else yet end users get exceptions to catch. I'm saying that as I'm currently doing a handler method that's doing if/else on the message text and I'd like to not do that.

**Comment by jknack** [comment]
Fixed (or at least more exceptions) in 2.0.0
